homeassistant/components/mqtt_eventstream.py

```diff
@@ -0,0 +1,93 @@
+"""
+homeassistant.components.mqtt_eventstream
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Connect multiple Home Assistant instances via mqtt.
+
+Key difference from using the "remote" component:
+* The Home Assistant instance(s) may be behind a firewall that
+  will not route HTTP traffic to it from external networks.
+* If a subscribing Home Assistant instance is unavailable, messages
+  for it will buffer on the MQTT queue for later delivery.
+
+Configuration:
+
+To use the mqtt_eventstream component you will need to add the component to
+your configuration.yaml file.
+
+If you do not specify a publish_topic you will not forward events to the queue.
+If you do not specify a subscribe_topic then you will not receive events from
+the remote server.
+
+mqtt_eventstream:
+  publish_topic: MyServerName
+  subscribe_topic: OtherHaServerName
+"""
+import json
+from homeassistant.core import EventOrigin, State
+from homeassistant.const import (
+    MATCH_ALL, EVENT_TIME_CHANGED, EVENT_CALL_SERVICE,
+    EVENT_SERVICE_EXECUTED, EVENT_STATE_CHANGED)
+import homeassistant.loader as loader
+from homeassistant.remote import JSONEncoder
+
+DOMAIN = "mqtt_eventstream"
+DEPENDENCIES = ['mqtt']
+
+
+def setup(hass, config):
+    """ Setup our mqtt_eventstream component. """
+    def _event_handler(event):
+        """ Handle events by publishing them on the mqtt queue. """
+        if event.origin != EventOrigin.local:
+            return
+        if event.event_type in (
+                EVENT_TIME_CHANGED,
+                EVENT_CALL_SERVICE,
```

Comment:
Ah yes, that is a tricky one. As this component is about event forwarding and very closely coupled to the MQTT component, it _might_ make sense to not use the publish service but use the internal method directly:

```python
from homeassistant.components import mqtt
mqtt.MQTT_CLIENT.publish(msg_topic, payload, qos, retain)
```

This will also prevent the history component from recording every event twice.

I am actually not sure if I like this because it goes against the core principle that the event bus is the history of everything happening in the system.